### PR TITLE
Introduce dataset management and datapackage operations

### PR DESCRIPTION
**Pull request by rflprr**
- Add wrapper methods for REST API
- Introduce support for datapackages
- Refactor query functions
- Consolidate package interface
- Introduce unit tests
- Introduce configuration tool

**IMPORTANT**: The code under `datadotworld/client/_swagger` is auto-generated. I strongly recommend not spending any time on those.

### CODE REVIEW

File: datadotworld/client/api.py [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+        summary : str, optional
+            Dataset summary markdown
+        tags : list, optional
+            Dataset tags
+        license : {'Public Domain', 'PDDL', 'CC-0', 'CC-BY', 'ODC-BY', 'CC-BY-SA', 'ODC-ODbL', 'CC BY-NC', 
+                   'CC BY-NC-SA', 'Other'}
+            Dataset license
+        visibility : {'OPEN', 'PRIVATE'}
+            Dataset visibility
+        files : dict
```
Comment (shawnsmith) [inline datadotworld/client/api.py]: What are the keys and values of the dict?  Is this `names -> urls`?  Should be optional, and there should be text explaining that the `files` dict is only relevant if the files-to-add are publicly available on the internet, direct upload is possible via a different api call.

File: datadotworld/client/api.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+
+        Examples
+        --------
+        >>> api_client = datadotworld.api_client()
+        >>> api_client.upload_files('jonloyens/an-intro-to-dataworld-dataset',
+        >>>                 ['/Users/jon/DataDotWorldBBall/DataDotWorldBBallTeam.csv'])
+        """
+        owner_id, dataset_id = parse_dataset_key(dataset_key)
+        try:
+            self._uploads_api.upload_files(owner_id, dataset_id, files)
```
Comment (shawnsmith) [inline datadotworld/client/api.py]: This passes a list of files but the uploads_api signature accepts a single file.

File: datadotworld/client/api.py [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+        """
+        owner_id, dataset_id = parse_dataset_key(dataset_key)
+        try:
+            self._datasets_api.delete_files_and_sync_sources(owner_id, dataset_id, names)
+        except _swagger.ApiException as e:
+            raise RestApiError(cause=e)
+
+    # Datapackage
+
+    def download_datapackage(self, dataset_key, dest_dir):
```
Comment (shawnsmith) [inline datadotworld/client/api.py]: Implicit assumption that `dest_dir` doesn't exist already?

File: datadotworld/datadotworld.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            'User-Agent': _user_agent(),
+            'Accept': 'text/csv',
+            'Authorization': 'Bearer {0}'.format(self._config.auth_token)
+        }
+        response = requests.get(url, params=params, headers=headers)
+        if response.status_code == 200:
+            return QueryResults(response.text)
+        raise RuntimeError('Error executing query: {}'.format(response.text))
+
+    def load_dataset(self, dataset_key, force_update=False):
```
Comment (shawnsmith) [inline datadotworld/datadotworld.py]: I'd like an optional argument to override the destination directory.  Doc that the default goes to `~/.dw/cache` (or the cache_dir setting in the config file, unless it's not configurable?).

Comment (rflprr) [inline datadotworld/datadotworld.py]: Just not configurable for now.

File: datadotworld/datadotworld.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        RestApiError
+            If a server error occurs
+        """
+        owner_id, dataset_id = parse_dataset_key(dataset_key)
+        cache_dir = path.join(self._config.cache_dir, owner_id, dataset_id, 'latest')
+
+        backup_dir = None
+        if path.isdir(cache_dir) and force_update:
+            backup_dir = path.join(self._config.cache_dir, owner_id, dataset_id, 'backup')
+            shutil.move(cache_dir, backup_dir)
```
Comment (rflprr) [inline datadotworld/datadotworld.py]: Good catch. Will delete `backup_dir` if one exists.

File: datadotworld/client/swagger-dwapi-def.json [hunk 6]
```diff
@@ -0,0 +1,1400 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Dataset created successfully.",
+            "schema": {
+              "$ref": "#/definitions/SuccessMessage"
+            }
+          },
+          "202": {
```
Comment (rflprr) [inline datadotworld/client/swagger-dwapi-def.json]: That change is inconsequential (doc aside). I'll re-generate the swagger client when we have new API functionality to use.

File: datadotworld/client/api.py [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+                                            header_value='Bearer {}'.format(self._config.auth_token))
+        swagger_client.user_agent = _user_agent()
+
+        self._datasets_api = _swagger.DatasetsApi(swagger_client)
+        self._uploads_api = _swagger.UploadsApi(swagger_client)
+
+    # Dataset Operations
+
+    def get_dataset(self, dataset_key):
+        """Retrieve an existing dataset
```
Comment (rflprr) [inline datadotworld/client/api.py]: Done.

File: datadotworld/client/api.py [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+
+    def create_dataset(self, owner_id, **kwargs):
+        """Create a new dataset
+
+        Parameters
+        ----------
+        owner_id : str
+            Username of the owner of the new dataset
+        title : str
+            Dataset title
```
Comment (rflprr) [inline datadotworld/client/api.py]: Done.

File: datadotworld/client/api.py (hunk 1 above)
Comment (rflprr) [inline datadotworld/client/api.py]: Documented as optional, now. Thanks.

File: datadotworld/client/api.py (hunk 3 above)
Comment (rflprr) [inline datadotworld/client/api.py]: Added exception for the case where it already exists.

File: datadotworld/models/dataset.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        # ``data`` will be returned as bytes.
+        upcast_resource = datapackage.Resource(
+            self.__resources[resource_name].descriptor,
+            default_base_path=self.__base_path)
+        return upcast_resource.data
+
+    def _to_dataframe(self, resource_name):
+        """Extract dataframe from tabular resource"""
+
+        from jsontableschema_pandas import Storage
```
Comment (jboutros) [inline datadotworld/models/dataset.py]: should do the check for existence of pandas here?

File: README.md [hunk 10]
```diff
@@ -11,99 +13,103 @@ you can install it using `pip` directly from this github repo:
...
+For example:
+```python
+intro_dataset = dw.load_dataset('jonloyens/an-intro-to-dataworld-dataset')
 ```
 
-### Querying
+Dataset objects allow access to data in 3 different ways, via `raw_data`, `tables` and `dataframes` properties.  
+For example:
+```python
+>>> list(intro.dataframes)
```
Comment (jboutros) [inline README.md]: you should continue to use `intro_dataset` for continuity in these examples

**Pull request merged**